src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,135 @@
+        if (size == elementList.length) {
+            elementList = grow();
+        }
+        System.arraycopy(elementList, index, elementList, index + 1, size - index);
+        elementList[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        Object[] listAsArray = list.toArray();
+        if (listAsArray.length > elementList.length - size) {
+            elementList = grow(listAsArray.length + size);
+        }
+        System.arraycopy(listAsArray, 0, elementList, size, listAsArray.length);
+        size += listAsArray.length;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index > size - 1 || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Can not get element, "
+                    + "index is out of bounds.");
+        }
+        return (T) elementList[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index > size - 1 || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Can not set element, "
+                    + "index is out of bounds.");
+        }
+        elementList[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index > size - 1 || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Can not remove element, "
+                    + "index is out of bounds.");
+        }
+        int newSize = size - 1;
+        final Object oldValue = elementList[index];
+        if (newSize > index) {
+            System.arraycopy(elementList, index + 1, elementList, index, newSize - index);
+        }
+        size = newSize;
+        elementList[size] = null;
+        return (T) oldValue;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int index = -1;
+        for (int i = 0; i < size; i++) {
+            if ((element == elementList[i])
+                    || (element != null && element.equals(elementList[i]))) {
+                index = i;
+                break;
+            }
+        }
+        if (index == -1) {
+            throw new NoSuchElementException("Element not found");
+        }
+        Object oldValue = elementList[index];
+        remove(index);
+        return (T) oldValue;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    @Override
+    public Object[] toArray() {
+        Object[] elementData = new Object[size];
+        System.arraycopy(elementList, 0, elementData, 0, size);
+        return elementData;
+    }
+
+    private Object[] grow(int minCapacity) {
+        int oldCapacity = elementList.length;
+        if (elementList != EMPTY_ELEMENT_DATA) {
+            int minGrowth = minCapacity - oldCapacity;
+            int defaultGrowth = oldCapacity >> 1;
```

Comment:
avoid bitwise operations, gains are insignificant, readability is more important, create a GROW_FACTOR constant